[docs] Add "Building design system components" guide with Pigment CSS (#41635)
mui/material-ui: packages/pigment-css-react/README.md

  )
}
```

> ## Building a UI library

Reviewer: ```suggestion
## Building a reusable component
```

Seems more fitting here? The idea is to teach how to build a component that can have a styling-related prop interface using Pigment, so I'd probably go somewhere this route!

Author: Using "UI library" is more appropriate for this section. "reusable component" is too broad, developers can create "reusable component" with a simple `styled('div')(…)`.

This section is specific to design system authors who want to build a library that will be shared among projects. cc @samuelsycamore to help on this.

Reviewer: > This section is specific to design system authors who want to build a library that will be shared among projects.

```suggestion
## Building a design system component
```

Something along these lines? I find the "Building a UI library" title weird because we're demonstrating how to build just one component, not all that you need for a library. But I see what you mean; it's like a "design system-ready component", more than a simple reusable one.

> The key of a UI library is to provide a set of reusable components that can be used and themed across **different projects**.

Reviewer: I'd probably mention that the approach we'll follow here (using slots) is meant for folks wanting to build a highly customizable component. If that's not the case (i.e., you're good with having the component a bit more closed off), then this could be a bit of overkill.

Author: Probably need @samuelsycamore advice here. My intention is to tell developers that you want to read this section if you are building a UI library that will be shared among projects. If you want to build components for just one project, this approach might be over-engineered.

> This guide will show you how to build a statistics component as part of a UI library using Pigment CSS. There are 3 steps to follow:

Reviewer: ```suggestion
This guide will show you how to build a KPI card component as part of a UI library using Pigment CSS. There are 3 steps to follow:
```

Trying to find a better name for this component 😃 Just a quick suggestion.

Author: Sounds good to me.

1. Create the component slots.
2. Compose slots to create the component.
3. Style the slot with `ownerState`.

### 1. Create the component slots

Slots let the consumers customize each individual element of the component by targeting its respective name in the [theme's styleOverrides](#themeable-statistics-component).

This statistics component is composed of three slots:

- `root`: the container of the component
- `value`: the number of the statistics
- `unit`: the unit or description of the statistics

> 💡 Though you can give these slots any names you prefer, we recommend using `root` for the outermost container element for consistency with the rest of the library.

Use the `styled` API with `name` and `slot` parameters to create the slots, as shown below:

```js
import * as React from 'react';
import { styled } from '@pigment-css/react';

const StatRoot = styled('div', {

>   name: 'MuiStat', // The component name

Reviewer: ```suggestion
  name: 'PigmentStat', // The component name
```

Would avoid using "Mui" — it could be whatever, I guess (just also not "Pig" 😬)

Author: 😅 To confess, I used `PigStat` before changed to `MuiState`.

  slot: 'root', // The slot name
})(({ theme }) => ({
  display: 'flex',
  flexDirection: 'column',
  gap: '1rem',
  padding: '0.75rem 1rem',
  backgroundColor: theme.colors.primary.background,
  borderRadius: theme.radius.xs,
  boxShadow: theme.shadow.sm,
  letterSpacing: '-0.025em',
  fontWeight: 600,
}));

const StatValue = styled('div', {
  name: 'MuiStat',
  slot: 'value',
})(({ theme }) => ({
  ...theme.typography.h3,
}));

const StatUnit = styled('div', {
  name: 'MuiStat',
  slot: 'unit',
})(({ theme }) => ({
  ...theme.typography.body2,
  color: theme.colors.neutral.foreground,
}));
```

### 2. Create the component

Assemble the component using the slots created in the previous step:

```js
// /path/to/Stat.js
import * as React from 'react';

const StatRoot = styled('div', {
  name: 'MuiStat',
  slot: 'root',
})(…);

const StatValue = styled('div', {
  name: 'MuiStat',
  slot: 'value',
})(…);

const StatUnit = styled('div', {
  name: 'MuiStat',
  slot: 'unit',

> })(…);

Reviewer 2: These are redundant here given they are defined just above.

Reviewer: Yeah, was just about to ask what's the different between this section and previous, too

const Stat = React.forwardRef(function Stat(props, ref) {
  const { value, unit, ...other } = props;

  return (
    <StatRoot ref={ref} {...other}>
      <StatValue>{value}</StatValue>
      <StatUnit>{unit}</StatUnit>
    </StatRoot>
  );
});

export default Stat;
```

### 3. Style the slot with ownerState

> When you need to style the slot-based props or internal state, wrap them in the `ownerState` object and pass it to each slot as a prop.

Author: I think `ownerState` is also Pigment specific. It's in all of the Pigment codebase. By using `ownerState`, you don't have to write `shouldForwardProp` all the time. It makes a lot of sense to me to mention `ownerState` for styling.

> We should also mention that users get access to the direct props as well.

Yes, we can do that.

> When you need to style the slot-based props or internal state, wrap them in the `ownerState` object and pass it to each slot as a prop.

Reviewer 2: It's added only to support material-ui. Otherwise, it's not needed and has nothing to do with Pigment itself.

> When you need to style the slot-based props or internal state, wrap them in the `ownerState` object and pass it to each slot as a prop.

Author: Okay, I won't mention `ownerState` for now but I feel that developers will have to come up with some name anyway, so why not use what we already have?

It's not just for Material UI but quite common use cases. I don't think anyone would like to deal with `shouldForwardProp`, it's pretty annoying. For me, I just need Pigment to tell me what prop name I should put info for styling into and that's done. I don't want to be aware of props spreading to DOM.

    name: 'MuiStat',
    slot: 'root',
-  })(({ theme }) => ({
+  })(({ theme, ownerState }) => ({
    display: 'flex',
    flexDirection: 'column',
    gap: '1rem',
    padding: '0.75rem 1rem',
    backgroundColor: theme.colors.primary.background,
    borderRadius: theme.radius.xs,
    boxShadow: theme.shadow.sm,
    letterSpacing: '-0.025em',
    fontWeight: 600,
+   variants: [
+     {
+       props: { variant: 'outlined' },
+       style: {
+         border: `2px solid #e9e9e9`,
+       },
+     },
+   ],
  }));
```

That's it! You've successfully created a statistics component with Pigment CSS. The next step is to upload your component to a package registry to let other people use it.

### Consumer usage

The developers who want to use your component need to install your package and Pigment packages based on their [framework](#start-with-nextjs).

```bash
npm install your-package-name @pigment-css/react
npm install -D @pigment-css/nextjs-plugin
```

Then, they need to set up Pigment CSS in their project and import the stylesheet in the root layout file:

```js
// framework config file, for example next.config.js
const { withPigment } = require('@pigment-css/nextjs-plugin');

> const theme = {

Reviewer 2: Maybe we can also emphasize that lub authors can also export their pre configured theme that can be directly imported and passed to the config.

Author: I decided not to mention the theme configuration at the beginning. I think the theme should be added after teaching how to create a component.

+        <StatValue ownerState={ownerState}>{value}</StatValue>
+        <StatUnit ownerState={ownerState}>{unit}</StatUnit>
+      </StatRoot>
    );
  });
```

Then you can read `ownerState` in the slot to style it based on the `variant` prop.

```diff
  const StatRoot = styled('div', {
    name: 'PigmentStat',
    slot: 'root',
  })({
    display: 'flex',
    flexDirection: 'column',
    gap: '1rem',
    padding: '0.75rem 1rem',
    backgroundColor: '#f9f9f9',
    borderRadius: '8px',
    boxShadow: '0 2px 4px rgba(0, 0, 0, 0.1)',
    letterSpacing: '-0.025em',
    fontWeight: 600,
+   variants: [
+     {
+       props: { variant: 'outlined' },
+       style: {
+         border: `2px solid #e9e9e9`,
+       },
+     },
+   ],
  });
```

That's it! You've successfully created a statistics component with Pigment CSS. The next step is to upload your component to a package registry to let other people use it.

### Consumer usage

The developers who want to use your component need to install your package and Pigment packages based on their [framework](#start-with-nextjs).

```bash
npm install your-package-name @pigment-css/react
npm install -D @pigment-css/nextjs-plugin
```

> Then, they need to set up Pigment CSS in their project:

Reviewer: Little typo here after "Next,", right? Should be "they" I think
